### ISSUE TITLE
 Comparison plots - Single vs. Array question 

### ISSUE THREAD

**Issue by lheckmann** [N2-Survey/SurveyFramework#77]
Produce plots for comparison for more complicated question type:

E.g. multiple choice questions compared to a single choice question
![B9 SQ001 _feelfree_holidays_hue](https://user-images.githubusercontent.com/74262873/136378662-54432b8a-132e-4b2e-8beb-e1199eebb8c0.png)

- [ ] Make a function that produces a comparison plot similar to the one above for array to single choice questions
or think of a more reasonable way of presenting the data

What should be adjustable:

- [ ] Color Scheme (with a default value that can be set in a header file)
- [ ] Displaying or hiding the question title
- [ ] Figure dimensions (but default set to something reasonable calculated from the number of answer options and distribution)
- [ ] Option for adding spaces between certain bars 
- [ ] Displaying the "No Answers" or hiding them
- [ ] Threshold for percentage to be printed out
- [ ] Wrapping option for text labels in case they get too long

**Comment by TheFlike** [comment]
This is an issue suited for beginners, since simple questions are not as complicated to combine with other answers. The only real necessary libraries would by numpy, pandas and matplotlib.pyplot as plt
I know that pandas can be quite usefull and seaborn can directly interact with pandas, but my experience is that the basic numpy and matplotlib.pyplot is less confusing for non programmers, especially if nothing else is thrown in the mix and they still feel a little bit shakey, so they only have to concentrate on two libraries. 

So my suggestion is to get as soon as possible the data you want to plot out of the pandas dataframe into simple arrays and lists, improvement can happen after a working script exists.

My start here would be dividing the arrayquestion into multiple plots. 
For the exemplary arrayquestion 'B6'  this would be:

s.get_responses('B6').columns
-> ['More time needed to complete PhD project', 'Parental leave',
       'Wrap-up phase after completion of the PhD project']

I would choose one of these columns and combine with 'A6' since it is the closest to an actual combination that could be examined in the N2Survey ('Have certain genders more options for paternal leave?' etc.)

So then I'd use:

plot_title = column
x=s.count('B6', labels=True)[column].index.values
x=x.astype(str)
combi_array = pd.concat(
        [
            s.get_responses(question[0], labels=True, drop_other=True)[column],
            s.get_responses(question[1], labels=True, drop_other=True),
        ],
        axis=1,
    ).values

here every participating person is assigned both their answers to both questions. If column='Parental leave' e.g. this results in
--> plot_title = 'Parental leave'
-->  x = array(['Yes', 'No', "I don't know","I don't want to answer this question", 'No Answer'], dtype='<U36')
-->combi_array= array([['Yes', 'Woman'],
       ['Yes', 'Woman'],
       ['Yes', 'Man'],
       ['Yes', "I don't want to answer this question"],
       ['No Answer', 'No Answer'],
       ["I don't know", 'Woman'],.....

now use np.unique:

(combinations, occurences) = np.unique(array.astype(str), return_counts=True, axis=0)

and you get all possible combinations for the bar lengths:
--> combinations = array([["I don't know", 'Man'],
       ["I don't know", 'Woman'],
       ['No', 'Woman'],
       ['No Answer', 'Gender diverse (Gender-fluid)'],
       ['No Answer', 'No Answer'],
       ['No Answer', 'Non-binary'],
       ['Yes', "I don't want to answer this question"],
       ['Yes', 'Man'],
       ['Yes', 'Non-binary'],
       ['Yes', 'Woman']], dtype='<U36')

and how often they occure
--> occurences = array([ 6,  4,  3,  1,  4,  1,  1,  3,  2, 11])

now you have simple data and a starting point for the further work with pyplot, which in itself is a still enough work, but it is so widely used that you would not have to search long for answers to problems if they occure.

If you have a working plot script with this data as Input save it in an extrafile in  n2survey/plot/ and we if you want we can continue from there.

Of course this is just a suggestion and if you want to use seaborn for plotting and pandas please feel free, just from my experience, pandas is closer to object oriented I think and if you are not a programmer but a scientist, chances are you'll feel more comfortable with simple arrays.

**Comment by lheckmann** [comment]
Nice, thank you!

Two things:
1. For the current version: would it maybe be more easy to see things when the color gradiant is in each bar for the likert scale and the hue (Woman, Men,...) is in different colors?

2. Can we also have a more "overview" way of showing thing. E.g. Summarizing the "satisfied" of "dissatisfied part" of the answers and then treating them the same way as for the multiple-single comp plots? The "satisfied part" can be defined by the used, stating e.g. that "Very satisfied", "satisfied" are the "satisfied part" and then they are aggregated and counted the same way as the "Yes" in the multiple choice?
Is it clear what I mean?

**Comment by TheFlike** [comment]
The second point works now, I set standard `overview= True`:
```
s.plot(
    "C1",
    compare_with="A6",
    save='array_single_ov.png',
    totalbar=True,
    threshold_percentage=0,
    suppress_answers = [
        "Does not apply",
        "I don’t want to answer this question"],
    legend_columns=2,
    legend_title = True,
    plot_title=True,
    bar_width = 0.8,
)
```
creates now:
![array_single_ov](https://user-images.githubusercontent.com/60603101/178225305-d40c7ada-c373-4132-9488-34725985b155.png)

while

```
s.plot(
    "C1",
    compare_with="A6",
    save='array_single.png',
    totalbar=True,
    threshold_percentage=0,
    suppress_answers = [],
    legend_columns=2,
    legend_title = True,
    plot_title=True,
    bar_width = 0.8,
    overview=False,
    
)
```
creates the more detailed version:

![array_single](https://user-images.githubusercontent.com/60603101/178225515-3306eeee-cd89-4b41-9cf2-de6ce3d0304a.png)

**Comment by TheFlike** [comment]
Standard: all neutral choices are grouped (e.g. here: does not apply, neither/nor, I don't want to answer), always.
If all neutral choices except one are suppressed via entry in `suppress_answers`, instead of "all neutral choices" it tells the leftover choice (e.g "neither/nor). 
`ignore_no_answer=True` is standard, if  `ignore_no_answer=False`  "No Answer" is counted as neutral choice.
`overview = True` is standard, should it stay that way?

The following code 
```
from n2survey import LimeSurvey
s = LimeSurvey("SurveyFramework/data/survey_structure_2021.xml")
s.read_responses("SurveyFramework/data/dummy_data_2021_codeonly.csv")
s.theme["rc"]["figure.figsize"] = (16, 9)
s.theme["palette"] = "Reds"
s.theme['font_scale']=1
s.plot(
    "C1",
    compare_with="A6",
    save='array_single.png',
    totalbar=True,
    threshold_percentage=0,
    suppress_answers = [
        "Supervision",],
    ignore_no_answer = False,
    legend_columns=2,
    legend_title = True,
    plot_title=True,
    bar_width = 0.8,
    overview = False
)
```
should create:
![array_single](https://user-images.githubusercontent.com/60603101/179363723-eb0d1aee-e625-49a1-b186-1bd329c532f0.png)

With standard `overview=True` and `ignore_no_answer = True` the code:
```
s.plot(
    "C1",
    compare_with="A6",
    save='array_single_ov.png',
    totalbar=True,
    threshold_percentage=0,
    suppress_answers = [
        "Supervision",],
    legend_columns=2,
    legend_title = True,
    plot_title=True,
    bar_width = 0.8,
    overview=True,
)
```
should create:
![array_single_ov](https://user-images.githubusercontent.com/60603101/179363845-aeafdd54-8774-4e88-9f51-f8aed9dec77d.png)

**Comment by TheFlike** [comment]
I also calculated some random bars by hand and the result is the same as in the plot, so I am quite confident that the calculation works.

**Comment by lheckmann** [comment]
First I'll state the bigger to do and later I'll check all of the smaller changes:

For the overview I did not mean using the same figure type but actually an easier one. Using the same one as for the single vs. multiple, where for the multiple you choose all "Yes" counts to calculate the bars. Now the idea would be to use all "positive" or all "negative" counts to calculate the bars. Does that make sense?

**Comment by TheFlike** [comment]
Maybe I don't understand it, but the bottom picture I uses all positive choices and combines them to a single "positive" choice, the same for negative.

**Comment by TheFlike** [comment]
Ooh, okay, so bar plots, with 3 Bars per comparison answer (positive negative and neutral)

**Comment by lheckmann** [comment]
No, only one bar. We neglect the other two and the use can choose if they want positive or negative. So it looks the same as the multiple vs. single plots.

**Comment by lheckmann** [comment]
And yes, the way you combine it is already fine :)

**Comment by TheFlike** [comment]
> Oh, okay, I think I got it, so ideally `overview=True `for the picture here, `overview='Positive'`, `overview = 'Negative'`, `overview='Neutral' ` for just the positive /negative/etc. bars and then there is only one bar to consider per single-choice answer.

Yes I can do that, so the pull request is not yet ready to merge.

**Comment by lheckmann** [comment]
Perfect, thank you! Let me know, when I should have a look again
